packages/@aws-cdk/aws-sns/lib/subscription.ts

```diff
@@ -36,7 +36,21 @@ export interface SubscriptionOptions {
    *
    * @default - all messages are delivered
    */
-  readonly filterPolicy?: { [attribute: string]: SubscriptionFilter };
+  readonly filterPolicy? : { [attribute: string]: SubscriptionFilter };
+
+  /**
+   * The filter policy. V2 is compatible with FilterPolicyScope set to "MessageBody"
+   *
+   * @default - all messages are delivered
+   */
+  readonly filterPolicyV2? : SubscriptionFilterPolicyV2;
+
+  /**
+   * The filter policy scope.
+   *
+   * @default -  Filter applied through message attributes
+   */
+  readonly filterPolicyScope?: SubscriptionFilterPolicyScope;
```

Comment:
I don't love the design of using `filterPolicyScope`. It would be simpler to just have one option for specifying a filterPolicy using the message attributes, and another option for a filterPolicy using the message body. What do you think about a design like this:

```
export interface SubscriptionOptions {
  ...
  // The filter policy is evaluated against the message attributes.
  readonly filterPolicy? : { [attribute: string]: SubscriptionFilter };

  // The filter policy is evaluated against the message body
  readonly messageBodyFilterPolicy? : SubscriptionFilterWithNesting
}

export interface SubscriptionFilterWithNesting {
  [attribute: string]: SubscriptionFilter | SubscriptionFilterWithNesting;
}
```

I like this better because
1. It removes the 'V2' suffixes from the API 
2. There are fewer properties that you need to specify or reason about
3. The message body filtering is actually a different type and concept than the message attribute filtering. Just because they are both modeled as a json object in the CFN API, doesn't mean we need to modify them both as the same type here.

Reply:
I was thinking this as an option too. I initially opted for the `filterPolicyScope` because it seemed like there was a one to one mapping of cloudformation keys and cdk construct props. However, I think it makes sense to just create something like `filterPolicyWithMessageBody`. That way if `filterPolicy` is non null, use `MessageAttributes` and if `filterPolicyWithMessageBody` is non null, use `MessageBody` (ie the filterPolicy itself will dictate which policy to use. I'll adjust that and make those changes in my next revision.